src/dstore/dstore_base.c

```diff
@@ -817,3 +773,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
 
 	return rc;
 }
+
+static int dstore_dealloc_op(struct dstore_obj *obj, struct dstore_io_vec *vec,
+			     struct dstore_io_op **out)
+{
+	int rc;
+
+	rc = dstore_io_op_init_and_submit(obj, vec, out,
+					  DSTORE_IO_OP_FREE);
+
+	log_debug(OBJ_ID_F " <=> %p, "
+		  "vec=%p *out=%p rc=%d",
+		  OBJ_ID_P(dstore_obj_id(obj)), obj, vec,
+		  rc == 0 ? *out : NULL, rc);
+
+	return rc;
+}
+
+static int dstore_deallocate(struct dstore_obj *obj, off_t offset, size_t count)
+{
+	int rc = 0;
+	char *tmp_buf = NULL;
+	struct dstore_io_op *dop = NULL;
+	struct dstore_io_vec vec;
+	size_t nr_request;
+	size_t ndata_per_req;
+	size_t tail_size;
+	int i;
+	size_t old_size;
+	size_t bsize;
+	/* For tracing purpose */
+	off_t off = offset;
+	size_t size = count;
+
+	dassert(obj);
+
+	bsize = dstore_get_bsize(obj->ds,
+				 (dstore_oid_t *)dstore_obj_id(obj));
+
+	/* calculate count again so as to make dealloc operation bsize aligned.
+	 * Example: if bsize is 4096 and file size is 3000, truncate to 0.
+	 * we can not deallocate 3000 bytes, we have to make operation bsize
+	 * aligned, else motr code will panic.
+	 */
+	old_size = count + offset;
+	if (old_size % bsize != 0) {
+		old_size = ((old_size + bsize) / bsize) * bsize;
+		count = old_size - offset;
+	}
+
+	if (offset % bsize != 0) {
+		/*we need to make deallocate operation left aligned */
+		tmp_buf = calloc(bsize, sizeof(char));
+		uint32_t left_blk_num = offset/bsize;
+		uint32_t write_count = offset - (left_blk_num * bsize);
+		write_count = bsize - write_count;
+
+		RC_WRAP_LABEL(rc, out, dstore_pwrite, obj,
+			      offset, write_count,
+			      bsize, tmp_buf);
+
+		if (write_count == count) {
+			goto out;
+		}
+
+		count = count - write_count;
+		offset = offset + write_count;
+	}
+
+	ndata_per_req = (DSAL_MAX_DEALLOC_OP_SIZE/bsize)*bsize;	
+	nr_request = count / ndata_per_req;
+	tail_size = count - (nr_request * ndata_per_req);
+
+	/* we do not need io buffer here, we just need to send extents info */
+	memset(&vec, 0, sizeof(struct dstore_io_vec));
```

Comment:
We can initialize it with zeros in the beginning: `struct dstore_io_vec vec = { .data = NULL };`.

Reply:
done.

Reply:
I am not able to do this, compile complains error: unknown field ‘data’ specified in initializer.
I do not wish to add each individual element init state there.

Reply:
sorry, I used the wrong name. It can be dbufs or any other field.

> I do not wish to add each individual element init state there.

It is not required. The C standard guarantees that the rest of the fields will be initialized with zeros:
```c
struct foo { int a; int *b; }
...
void f() {
 struct foo foo = { .a = 0 };
 assert(foo.b == NULL);
}
```

Anyway, if you believe memset looks ok here then let's keep using memset -- it is totally fine, we don't have a special coding style rule for that.